### PR TITLE
Refactor mounter

### PR DESCRIPTION
**Pull request by mossprescott**
Defines a `PathStore` trait with the necessary operations, with an implementation for `KeyValueStore.

### CODE REVIEW

File: core/src/main/scala/quasar/fs/mount/Mounter.scala [hunk 1]
```diff
@@ -123,23 +114,23 @@ object Mounter {
       case Unmount(path) =>
         handleUnmount(path)
           .flatMapF(_ =>
-            refineType(path).swap.foldMap(forPrefix)
+            refineType(path).swap.fold[F[Set[APath]]](κ(Set.empty.point[F]), store.descendants)
```
Comment (wemrysi) [inline core/src/main/scala/quasar/fs/mount/Mounter.scala]: I think you could use `foldMapM` here.

File: core/src/main/scala/quasar/fs/mount/Mounter.scala [hunk 2]
```diff
@@ -28,91 +28,82 @@ import scalaz._, Scalaz._
...
-  def apply[F[_], S[_]](
-    mount: MountRequest => F[MountingError \/ Unit],
-    unmount: MountRequest => F[Unit]
-  )(implicit
-    S0: F :<: S,
-    S1: MountConfigs :<: S
-  ): Mounting ~> Free[S, ?] = {
+  /** A kind of key-value store where the keys are absolute paths, with an
+    * additional operation to efficiently look up nested paths. */
+  trait PathStore[F[_], V] {
```
Comment (wemrysi) [inline core/src/main/scala/quasar/fs/mount/Mounter.scala]: Mind adding some documentation for the methods, especially `insert` and `recover` (i.e. what the `Boolean` represents and what the semantics of `recover` is)?

**Comment by wemrysi** [comment]
This looks good in general, I think the `recover` function is a bit smelly, though I don't have a great alternative (maybe a more declarative way of specifying an atomic region `atomic[A](fa: F[A]): F[A]`?).

**Comment by mossprescott** [comment]
I can't argue with your assessment of `recover`. In fact, I can't say for sure that it's even being used correctly because the tests passed before it was implemented and I only included it in the hope of preserving the previous behavior. I believe it may be necessary only in the case of concurrent conflicting `mount` operations, but looking at it again I'm not sure I understand how it would be necessary to rollback the transaction because it doesn't seem like any subsequent operations are done on `store`. It may depend on how the resulting program (containing `ConnectionIO`) is interpreted.

File: core/src/main/scala/quasar/fs/mount/Mounter.scala (hunk 2 above)
Comment (mossprescott) [inline core/src/main/scala/quasar/fs/mount/Mounter.scala]: I will. Alternatively, I could change the signature to `insert(path, value): MntErrT[F, Unit]` which is probably somewhat more self-explanatory. But I think I'll just add a comment.

**Comment by mossprescott** [comment]
I'm not 100% sure, but I think I can eliminate `recover` and just put the rollback directly into the `insert` action. Maybe we can look at it together tomorrow and decide if that seems correct.

**Comment by wemrysi** [comment]
@mossprescott I _think_ (if memory serves) that the `rollback` in QAdv was due to how the PostgreSQL java lib dealt with constraint violations. We are relying on the `PRIMARY KEY` constraint in the `Mounts` table to tell us if a mount already exists, but I think it puts the `Connection` in a state where every subsequent op fails with an error that the transaction is closed (or something like that), so we `rollback` to abort that transaction and start a new one, does that sound familiar?

If that is the case, then I think you're right that the `rollback` can be moved into the `insert` impl for the `MetaStoreMounter`.

**Pull request merged**